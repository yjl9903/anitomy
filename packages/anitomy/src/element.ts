export enum ElementCategory {
  AnimeSeason = 'season',
  AnimeSeasonPrefix = 'prefix.season',
  AnimeTitle = 'title',
  AnimeType = 'type',
  AnimeYear = 'year',
  AnimeMonth = 'month', // Added by https://github.com/yjl9903/AnimeGarden
  DeviceCompatibility = 'DeviceCompatibility',
  Source = 'source',
  // Episode
  EpisodeNumber = 'episode.number',
  EpisodeNumberAlt = 'episode.numberAlt',
  EpisodePrefix = 'prefix.episode',
  EpisodeTitle = 'episode.title',
  // File
  FileChecksum = 'checksum',
  FileExtension = 'extension',
  FileName = 'filename',
  // Language
  Language = 'language',
  Subtitles = 'subtitles',
  // Audio and Video
  AudioTerm = 'audio.term',
  VideoResolution = 'video.resolution',
  VideoTerm = 'video.term',
  // Volume
  VolumeNumber = 'volume',
  VolumePrefix = 'prefix.volume',
  // Release
  ReleaseGroup = 'release.group',
  ReleaseInformation = 'release.information',
  ReleaseVersion = 'release.version',
  // Other
  Unknown = 'unknown',
  Other = 'other'
}
