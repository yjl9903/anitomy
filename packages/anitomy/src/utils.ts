import type { ParsedResult } from './types';

export function inRange<T>(list: T[], idx: number) {
  return 0 <= idx && idx < list.length;
}

export function isNumericString(text: string) {
  return /^\d+$/.test(text);
}

export function trim(text: string, removal: string[]) {
  let start = 0,
    end = text.length - 1;
  while (start <= end && removal.includes(text[start])) {
    start++;
  }
  while (end >= start && removal.includes(text[end])) {
    end--;
  }
  return text.slice(start, end + 1);
}

export function mergeResult(source: ParsedResult, income: ParsedResult = {}) {
  return {
    ...source,
    ...income
  };
}
