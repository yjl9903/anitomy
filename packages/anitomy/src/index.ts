import type { AnitomyOptions, AnitomyResult, ParsedResult } from './types';

import { mergeResult } from './utils';
import { KeywordManager } from './keyword';
import { ElementCategory } from './element';
import { parse as doParse } from './parser';
import { tokenize as doTokenize } from './tokenizer';

export type * from './types';

/**
 * Parser instance with shared options and parsed result cache
 */
export class Parser {
  private readonly options: AnitomyOptions;

  private readonly cache = new Map<string, AnitomyResult | null>();

  public constructor(options: Partial<AnitomyOptions> = {}) {
    this.options = resolveOptions(options);
  }

  public parse(filename: string, force = false) {
    if (!force && this.cache.has(filename)) {
      const result = this.cache.get(filename);
      return result ? result : undefined;
    } else {
      const result = parse(filename, this.options);
      if (result) {
        this.cache.set(filename, result);
      } else {
        this.cache.set(filename, null);
      }
      return result;
    }
  }
}

export function parse(
  filename: string,
  _options: Partial<AnitomyOptions> = {}
): AnitomyResult | undefined {
  if (filename === '') return undefined;

  let result: ParsedResult = {};

  const options = resolveOptions(_options);

  result.filename = filename;
  if (options.parseFileExtension) {
    const ext = removeExtension(filename);
    if (ext) {
      result.filename = filename;
      result.extension = ext.extension;
    }
  }

  const tokenized = doTokenize(result.filename!, options);
  result = mergeResult(result, tokenized.result);
  if (!tokenized.ok) {
    return resolveResult(result);
  }

  const parsed = doParse(result, tokenized.tokens, options);
  result = parsed.result;

  return resolveResult(result);
}

export function resolveOptions(options: Partial<AnitomyOptions>): AnitomyOptions {
  return {
    delimiters: ' _.&+,|',
    parseEpisodeNumber: true,
    parseEpisodeTitle: true,
    parseFileExtension: true,
    parseReleaseGroup: true,
    ...options
  };
}

function resolveResult(result: ParsedResult): AnitomyResult {
  const resolved: AnitomyResult = {
    title: result['title'],
    type: result['type'],
    season: normalizeSeason(result['season']),
    year: normalizeNumber(result['year']),
    month: normalizeNumber(result['month']),
    language: result['language'],
    subtitles: result['subtitles'],
    source: result['source'],
    episode: {
      number: normalizeNumber(result['episode.number']),
      numberAlt: normalizeNumber(result['episode.numberAlt']),
      title: result['episode.title']
    },
    volume: {
      number: normalizeNumber(result['volume'])
    },
    video: {
      term: result['video.term'],
      resolution: result['video.resolution']
    },
    audio: {
      term: result['audio.term']
    },
    release: {
      version: normalizeNumber(result['release.version']),
      group: result['release.group']
    },
    file: {
      name: result['filename']!,
      extension: result['extension'],
      checksum: result['checksum']
    }
  };
  return resolved;

  function normalizeSeason(num: string | undefined) {
    if (num !== undefined && num !== null) {
      return /^\d+$/.test(num) ? String(+num) : num;
    } else {
      return undefined;
    }
  }

  function normalizeNumber(num: string | undefined) {
    try {
      if (num !== undefined && num !== null) {
        const n = parseFloat(num);
        return !Number.isNaN(n) ? n : undefined;
      }
    } catch {}
    return undefined;
  }
}

function removeExtension(filename: string) {
  const position = filename.lastIndexOf('.');
  if (position === -1) return undefined;

  const extension = filename.slice(position + 1);
  if (extension.length > 4 || !/^[a-zA-Z0-0]+$/.test(extension)) {
    return undefined;
  }

  if (!KeywordManager.contains(ElementCategory.FileExtension, extension)) {
    return undefined;
  }

  return {
    filename: filename.slice(0, position),
    extension
  };
}
