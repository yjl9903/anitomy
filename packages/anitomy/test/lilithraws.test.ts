import { describe, it, expect } from 'vitest';

import { parse } from '../src';

import { LilithRaws } from './testcase';

describe('Lilith-Raws', () => {
  it('should parse', () => {
    for (const filename of LilithRaws) {
      const info = parse(filename);
      expect(info).toMatchSnapshot();
      expect(info?.title).toBeTruthy();
    }
  });
});
