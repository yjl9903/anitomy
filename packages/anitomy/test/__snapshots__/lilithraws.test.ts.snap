// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lilith-Raws > should parse 1`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 3,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] MIX 第二季 ~第二個夏天，邁向晴空~ / MIX S02 - 03 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "MIX 第二季 ~第二個夏天，邁向晴空~ / MIX S02",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 2`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 江戶前精靈 / Edomae Elf - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "江戶前精靈 / Edomae Elf",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 3`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 肌肉魔法使 / Mashle - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "肌肉魔法使 / Mashle",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 4`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 小鳥之翼 / Birdie Wing - Golf Girls Story S02 - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "小鳥之翼 / Birdie Wing - Golf Girls Story S02",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 5`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 國王排名 勇氣的寶箱 / Ousama Ranking - Yuuki no Takarabako - 01 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "國王排名 勇氣的寶箱 / Ousama Ranking - Yuuki no Takarabako",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 6`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 勇者死了！ / Yuusha ga Shinda! - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "勇者死了！ / Yuusha ga Shinda!",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 7`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] Opus.COLORs 色彩高校星 / Opus.COLORs - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "Opus COLORs 色彩高校星 / Opus COLORs",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 8`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 百合是我的工作！ / Watashi no Yuri wa Oshigoto desu! - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "百合是我的工作！ / Watashi no Yuri wa Oshigoto desu!",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 9`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] Dr.STONE 新石紀 / Dr. Stone S03 - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "3",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "Dr STONE 新石紀 / Dr  Stone S03",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 10`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 魔法使的新娘 / Mahoutsukai no Yome S02 - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "魔法使的新娘 / Mahoutsukai no Yome S02",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 11`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 為美好的世界獻上爆焰！ / Kono Subarashii Sekai ni Bakuen wo! - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "為美好的世界獻上爆焰！ / Kono Subarashii Sekai ni Bakuen wo!",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 12`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 11,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 怕痛的我，把防禦力點滿就對了 / Bofuri S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "怕痛的我，把防禦力點滿就對了 / Bofuri S02",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 13`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 【我推的孩子】 / Oshi no Ko - 01 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "【我推的孩子】 / Oshi no Ko",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 14`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 魔術士歐菲 流浪之旅 聖域篇 / Majutsushi Orphen Hagure Tabi S04 Seiiki Hen - 01 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "4",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "魔術士歐菲 流浪之旅 聖域篇 / Majutsushi Orphen Hagure Tabi S04 Seiiki Hen",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 15`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 第二次被異世界召喚 / Isekai Shoukan wa Nidome desu - 01 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "第二次被異世界召喚 / Isekai Shoukan wa Nidome desu",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 16`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 絆之 Allele / Kizuna no Allele - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "絆之 Allele / Kizuna no Allele",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 17`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 在異世界獲得超強能力的我，在現實世界照樣無敵～等級提升改變人生命運～ / Iseleve - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "在異世界獲得超強能力的我，在現實世界照樣無敵～等級提升改變人生命運～ / Iseleve",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 18`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 屍體如山的死亡遊戲 / Dead Mount Death Play - 01 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "屍體如山的死亡遊戲 / Dead Mount Death Play",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 19`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 帶著智慧型手機闖蕩異世界 / Isekai wa Smartphone to Tomo ni S02 - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "帶著智慧型手機闖蕩異世界 / Isekai wa Smartphone to Tomo ni S02",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 20`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 熊熊勇闖異世界 PUNCH！ / Kuma Kuma Kuma Bear S02 - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "熊熊勇闖異世界 PUNCH！ / Kuma Kuma Kuma Bear S02",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 21`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 她去公爵家的理由 / Kanojo ga Koushaku-tei ni Itta Riyuu - 01 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "她去公爵家的理由 / Kanojo ga Koushaku-tei ni Itta Riyuu",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 22`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 機戰少女 Alice Expansion / Alice Gear Aegis Expansion - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "機戰少女 Alice Expansion / Alice Gear Aegis Expansion",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 23`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 鬼滅之刃 刀匠村篇 / Kimetsu no Yaiba - Katanakaji no Sato-Hen - 01 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "鬼滅之刃 刀匠村篇 / Kimetsu no Yaiba - Katanakaji no Sato-Hen",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 24`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 18,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 我與機器子 / Boku to Roboko - 18 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "我與機器子 / Boku to Roboko",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 25`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 我家的英雄 / My Home Hero - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "我家的英雄 / My Home Hero",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 26`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 轉生貴族的異世界冒險錄～不知自重的眾神使徒～ / Tensei Kizoku no Isekai Boukenroku - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "轉生貴族的異世界冒險錄～不知自重的眾神使徒～ / Tensei Kizoku no Isekai Boukenroku",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 27`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 機動戰士鋼彈 水星的魔女 / Mobile Suit Gundam：The Witch from Mercury S02 - 01 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "機動戰士鋼彈 水星的魔女 / Mobile Suit Gundam：The Witch from Mercury S02",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 28`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 鄰人似銀河 / Otonari ni Ginga - 01 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "鄰人似銀河 / Otonari ni Ginga",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 29`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 和山田談場 Lv999 的戀愛 / Yamada-kun to Lv999 no Koi wo Suru - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "和山田談場 Lv999 的戀愛 / Yamada-kun to Lv999 no Koi wo Suru",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Lilith-Raws > should parse 30`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 11,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] UniteUp! 眾星齊聚 / UniteUp! - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "UniteUp! 眾星齊聚 / UniteUp!",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;
